File: scripts/scratch.py
from web3 import Web3
from dotenv import load_dotenv
import time
import json
import os
import requests
import sys


load_dotenv()

node = os.getenv('GOERLI_URL')
web3 = Web3(Web3.HTTPProvider(node))


def get_oracle():
    response = requests.get('https://alpha.horse.link/api/config')
    data = response.json()
    return data['addresses']['marketOracle']

def load_oracle():
    address = get_oracle()
    response = requests.get('https://raw.githubusercontent.com/horse-link/contracts.horse.link/main/deployments/goerli/MarketOracle.json')
    data = response.json()
    abi = data['abi']
    contract = web3.eth.contract(address=address, abi=abi)
    return contract

def set_scratch(oracle, marketId, propositionId, odds, totalOdds, signature) -> None:
    try:
        # Can be any account with funds
        account_from = {
            'private_key': os.getenv('SETTLE_PRIVATE_KEY'),
            'address': '0xF33b9A4efA380Df3B435f755DD2C2AF7fE53C2d1',
        }

        print(signature)

        # Convert the signature to a tuple
        signature_tuple = [signature['v'], signature['r'], signature['s']]
        print("***** Signature tuple: ", signature_tuple)

        encodedProposition = propositionId.encode('utf-8')
        proposition_id = bytearray(encodedProposition)
        encodedMarket = marketId.encode('utf-8')
        
        market_id = bytearray(encodedMarket)

        # remove the first 2 characters of the market_id and proposition_id
        #market_id = market_id[2:]
        #proposition_id = proposition_id[2:]

        print(market_id)
        print(proposition_id)
        print(odds)
        
        tx = oracle.functions.setScratchedResult(market_id, proposition_id, int(odds), int(totalOdds), signature_tuple).buildTransaction(
            {
                'from': account_from['address'],
                'nonce': web3.eth.get_transaction_count(account_from['address']),
            }
        )

        tx_create = web3.eth.account.sign_transaction(
            tx, account_from['private_key'])

        tx_hash = web3.eth.send_raw_transaction(tx_create.rawTransaction)
        tx_receipt = web3.eth.wait_for_transaction_receipt(tx_hash)

        print(tx_receipt)
    except:
        print("An exception occurred")
        #output the error
        print(sys.exc_info())

def hydrate_market_id(market_id):
    # Remove the '0x' prefix
    market_id = market_id[2:]
    
    # Convert hexadecimal to binary
    binary = bytes.fromhex(market_id)
    # Decode binary as ASCII
    market_string = binary.decode("ascii")
 
    # Parse the market string
    id = market_string[0:11]
    date = int(market_string[0:6])
    location = market_string[6:9]
    race = int(market_string[9:11])
    # Return a dictionary with the hydrated market data
    result = {
        "id": str(id),
        "date": date,
        "location": location,
        "race": race
    }
    return result


def get_subgraph_bets_since(createdAt_gt):
    # We may want to generate the schema locally with:
    # gql-cli 'https://api.thegraph.com/subgraphs/name/horse-link/hl-protocol-goerli' --print-schema > schema.graphql

    # From https://thegraph.com/hosted-service/subgraph/horse-link/hl-protocol-goerli
    bets_query = """
    {
      bets(where: {createdAt_gt: %(createdAt_gt)s},orderBy:createdAt, first: 1000) {
        id
        createdAt
        createdAtTx
        marketId
      }
    }
    """
    query = bets_query % {"createdAt_gt": createdAt_gt}
    print(query)
    response = requests.post(
        "https://api.thegraph.com/subgraphs/name/horse-link/hl-protocol-goerli",
        data=json.dumps({"query": query}),
    )
    data = json.loads(response.text)["data"]
    return data["bets"]


def main():
    print("Main")
  
    try:
      with open("state.json", "r") as state_file:
        state = json.load(state_file)
    except FileNotFoundError:
      state = {"last_run": 0, "watch_list": [], "processed_items": []}
    oracle = load_oracle()
    print("Loaded")
    print(f"Watch list contains {len(state['watch_list'])} races")

    # Fetch new bets from subgraph
    last_run = int(state.get("last_run", 0))
    this_run = int(time.time())
    bets = get_subgraph_bets_since(last_run)
    print(f"Found {len(bets)} new bets")
    new_markets_count = 0
    for bet in bets:
      market_id = bet.get("marketId")
      if market_id not in state.get("watch_list", []):
        state["watch_list"].append(market_id)
        new_markets_count += 1

    print(f"Added {new_markets_count} new markets to watch list")
        
    # For each market in the watch list, fetch the race details
    processed_items = [] # array of propositions that have already been sent to the Oracle contract
    for market_id in state.get("watch_list", []):
      # hydrate market ID
      hydrated_market = hydrate_market_id(market_id)
      location = hydrated_market["location"]
      race = hydrated_market["race"]
      id = hydrated_market["id"]

      print(f"Processing race {location} {race}")
      market_result_url = "https://alpha.horse.link/api/bets/sign/" + id
      print(market_result_url)
      market_response = requests.get(market_result_url)
      market_data = json.loads(market_response.text)
      
      
      # Process scratched runners
      runners = market_data.get("scratchedRunners")
      if runners is None:
        print("No scratched runners")
        continue
      print(f"Found {len(runners)} scratched runners")

      for runner in runners:

        print("Processing scratched runner")
        print(market_data)
        proposition_id = runner["b16propositionId"]
        # If not already processed,
        if proposition_id not in processed_items:
          # Send this proposition to Oracle contract
          print(f"Sending proposition {proposition_id} to Oracle contract")
          odds = runner["odds"] * 1000000
          totalOdds = int(runner["totalOdds"])
          signature = runner["signature"]
          print(f"set_scratch(oracle, {proposition_id}, {market_id}, {odds}, {totalOdds}, {signature}")
          set_scratch(oracle, proposition_id, market_id, odds, totalOdds, signature)
          print("Sent.")

          # Add to processed_items
          processed_items.append(proposition_id)

      if market_data.get("status") == "closed":
        state["watch_list"].remove(market_id)

      state["last_run"] = this_run

    print("Saving state")
    with open("state.json", "w") as state_file:
      json.dump(state, state_file)
    print("Done")

if __name__ == "__main__":
    print(f"Starting {__file__} script at {int(time.time())}")
    main()

File: scripts/scratch.ts
import { ethers } from "ethers";
import axios from "axios";
import * as fs from "fs";
import * as dotenv from "dotenv";
import { getSubgraphBetsSince, loadOracle } from "./utils";
import type { Signature } from "./utils";
import { MarketDetails, formatting, markets } from "horselink-sdk";

// load .env into process.env
dotenv.config();

const baseApiUrl = process.env.API_URL ?? "https://alpha.horse.link/api";
console.log(`Using API URL: ${baseApiUrl}`);

async function setScratch(
	oracle: ethers.Contract,
	marketId: string,
	propositionId: string,
	odds: string,
	totalOdds: string,
	signature: Signature
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
): Promise<any> {
	const bnOdds = ethers.utils.parseUnits(Number(odds).toFixed(6), 6);
	const bnTotalOdds = ethers.utils.parseUnits(Number(totalOdds).toFixed(6));
	const encodedProposition = formatting.formatBytes16String(propositionId);
	const encodedMarket = formatting.formatBytes16String(marketId);

	try {
		const receipt = await oracle.setScratchedResult(
			encodedMarket,
			encodedProposition,
			bnOdds,
			bnTotalOdds,
			signature
		);
		return receipt;
	} catch (err) {
		console.error(err);
		return;
	}
}

async function main() {
	let state;
	try {
		state = JSON.parse(fs.readFileSync("state.json", "utf8"));
	} catch (error) {
		console.log("No state file found, starting from scratch");
		state = { last_run: 0, watch_list: [], processed_items: [] };
	}
	const oracle = await loadOracle();
	console.log(`Watch list contains ${state.watch_list.length} races`);

	// Fetch bets placed since the last run. Add their market IDs to the watch list.
	const lastRun = state.last_run ?? 0;
	const thisRun = Math.floor(Date.now() / 1000);
	const bets = await getSubgraphBetsSince(lastRun);
	console.log(`Found ${bets.length} new bets`);
	let newMarketsCount = 0;
	for (const bet of bets) {
		const marketId = formatting.bytes16HexToString(bet.marketId);
		if (!state.watch_list.includes(marketId)) {
			state.watch_list.push(marketId);
			newMarketsCount++;
		}
	}
	console.log(`Added ${newMarketsCount} new markets to watch list`);

	// For each market in the watch list, fetch the race details
	const processedItems: string[] = [];
	for (const marketId of state.watch_list) {
		// hydrate market ID
		const hydratedMarket = markets.rehydrateMarketId(marketId);
		const location = hydratedMarket.location;
		const race = hydratedMarket.race;

		// Get the scratch data from the API
		const marketResultUrl = `${baseApiUrl}/bets/${marketId}?sign=true`;
		const marketResponse = await axios.get(marketResultUrl);
		const marketData = marketResponse.data;

		// Process any scratched runners
		const runners = marketData.scratchedRunners;
		console.log(
			`Market ${marketId} has ${runners?.length ?? "no"} scratched runners`
		);
		if (runners && runners.length > 0) {
			console.log("======================================");
			console.log(
				`Venue: ${location} Race: ${race} Date: ${new Date(
					hydratedMarket.date * 24 * 60 * 60 * 1000
				)}`
			);
			console.log(`${runners.length} scratched runners`);
			for (const runner of runners) {
				try {
					const propositionId = formatting.bytes16HexToString(
						runner.b16propositionId
					);
					const proposition = markets.hydratePropositionId(propositionId);
					console.log(`  - ${proposition.number}`);
					// If not already processed,
					if (!processedItems.includes(propositionId)) {
						if (!runner.signature) {
							console.log("    already registered");
							continue;
						}

						// Send this proposition to Oracle contract
						console.log(`    sending ${propositionId} to Oracle`);
						const signature = runner.signature;

						const receipt = await setScratch(
							oracle,
							marketId,
							propositionId,
							runner.odds,
							runner.totalOdds,
							signature
						);
						console.log("    tx hash:", receipt.hash);

						// Add to processed_items
						processedItems.push(propositionId);
					}
				} catch (error) {
					console.log("    Error processing scratched runner:", error);
				}
			}
		}
		if (marketData.marketResultAdded) {
			console.log(`Removing closed market ${marketId} from watch list`);
			state.watch_list = state.watch_list.filter((id) => id !== marketId);
		}
	}
	state.last_run = thisRun;
	console.log("Writing state");
	fs.writeFileSync("./state.json", JSON.stringify(state, null, 4));
}

main()
	.then(() => {
		console.log("Done");
		process.exit(0);
	})
	.catch((error) => {
		console.error(error);
		process.exit(1);
	});
